internal/kubernetes/drainer.go

```diff
@@ -161,6 +170,68 @@ func (d *APICordonDrainer) Cordon(n *core.Node) error {
 	return nil
 }
 
+// MarkDrain set a condition on the node to mark that that drain is scheduled.
+func (d *APICordonDrainer) MarkDrain(n *core.Node, when, finish time.Time, failed bool) error {
+	nodeName := n.Name
+	// Refresh the node object
+	freshNode, err := d.c.CoreV1().Nodes().Get(nodeName, meta.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	msgSuffix := ""
+	conditionStatus := core.ConditionTrue
+	if !finish.IsZero() {
+		if failed {
+			msgSuffix = fmt.Sprintf(" | Failed: %s", finish.Format(time.RFC3339))
+		} else {
+			msgSuffix = fmt.Sprintf(" | Completed: %s", finish.Format(time.RFC3339))
+		}
+		conditionStatus = core.ConditionFalse
+	}
+
+	// Create or update the condition associated to the monitor
+	now := meta.Time{Time: time.Now()}
+	conditionUpdated := false
+	for i, condition := range freshNode.Status.Conditions {
```

Comment:
I'm just cross referencing this implementation with e.g. how the [NodeProblemDetector](https://github.com/kubernetes/node-problem-detector/blob/master/pkg/exporters/k8sexporter/problemclient/problem_client.go#L129) does it.

It looks like they're using the rest patch to use a strategic merge to update conditions. This kind of approach may be safer in the case of concurrent updates.

Reply:
Reviewing https://github.com/kubernetes/kubernetes/blob/release-1.1/docs/devel/api-conventions.md#concurrency-control-and-consistency this should be okay with the optimistic concurrency controls in place. 👍